src/app/beer_garden/api/http/authentication/login_handlers/ldap.py

```diff
@@ -0,0 +1,122 @@
+import logging
+from datetime import datetime, timezone
+from typing import Optional
+
+from brewtils.models import User
+from ldap3 import Connection, Server
+from ldap3.core.exceptions import LDAPException
+from mongoengine import DoesNotExist
+from tornado.httputil import HTTPServerRequest
+
+from beer_garden import config
+from beer_garden.api.http.authentication.login_handlers.base import BaseLoginHandler
+from beer_garden.api.http.schemas.v1.token import TokenInputSchema
+from beer_garden.role import get_role
+from beer_garden.user import create_user, get_user, update_user
+
+logger = logging.getLogger(__name__)
+
+
+class LdapLoginHandler(BaseLoginHandler):
+    """Handler for username and password ldap based authentication"""
+
+    @staticmethod
+    def get_user_dn(username: str):
+        """This combines user information into a complete user DN"""
+        if config.get("ldap.use_full_user_dn"):
+            return username
+        else:
+            dn_parts = (
+                f"{config.get('ldap.user_prefix')}={username}",
+                config.get("ldap.user_attributes"),
+                config.get("ldap.base_dn"),
+            )
+            return ",".join([s for s in dn_parts if s])
+
+    def verify_ldap_password(self, conn: Connection):
+        """Checks the provided plaintext password against the user's stored password
+
+        Args:
+            password: Plaintext string to check against user's password"
+
+        Returns:
+            bool: True if the password matches, False otherwise
+        """
+        if conn.result["description"] == "success":
+            return True
+        return False
+
+    def get_user_roles(self, conn: Connection, username: str):
+        """Checks the users roles against the provided"""
+        groups = set(config.get("ldap.default_user_roles"))
+        roles = []
+        conn.search(
+            config.get("ldap.roles_search_base"),
+            f"(&(objectclass=groupOfNames)(member={self.get_user_dn(username)}))",
+            attributes=["cn"],
+        )
+        for entry in conn.entries:
+            groups.add(entry["cn"].value)
+
+        for group in groups:
+            try:
+                roles.append(get_role(role_name=group))
+            except DoesNotExist:
+                pass
+
+        logger.info(
+            f"Updating {username} upstream roles to {[role.name for role in roles]}"
+        )
+        return roles
+
+    def get_user(self, request: HTTPServerRequest) -> Optional[User]:
+        """Gets the User corresponding to the username and password supplied in the
+        request body
+
+        Args:
+            request: tornado HTTPServerRequest object
+
+        Returns:
+            User: The User object matching the supplied username and password
+            None: If no User was found or the supplied password was invalid
+        """
+        authenticated_user = None
+
+        if request.body:
```

Comment:
We should check if it is enabled before we do any checks. Probably should do this in Trusted Headers as well.

`if config.get("ldap.enabled") and request.body:`

Reply:
There is already an enabled_login_handlers() function so this seems a bit redundant. What do you think? https://github.com/beer-garden/beer-garden/blob/438ab0a2c716d851747ce359644fab57dce2b609/src/app/beer_garden/api/http/authentication/login_handlers/__init__.py#L13C5-L13C29